theano/tensor/basic.py

```diff
@@ -2588,6 +2588,175 @@ def nonzero_values(a):
+            out[i][0] = res[i]
+
+
+def unravel_index(indices, dims, order='C', ndim=None):
+    """
+    Converts a flat index or array of flat indices into a tuple
+    of coordinate arrays.
+
+    This method is similar to the NumPy version, except for the
+    additional ``ndim`` parameter. This parameter is required if
+    the length of ``dims`` cannot be determined automatically, for
+    example because ``dims`` is a Theano vector.
+
+    For example:
+
+    >>> x = T.matrix()
+    >>> idx = T.ivector()
+    >>> unraveled_idx = T.unravel_index(idx, x.shape, ndim=x.ndim)
+
+    Parameters
+    ----------
+    indices : Theano or NumPy array
+        An integer array whose elements are indices into the flattened
+        version of an array of dimensions ``dims``.
+    dims : tuple of ints
+        The shape of the array to use for unraveling ``indices``.
+    order : {'C', 'F'}, optional
+        Determines whether the indices should be viewed as indexing in
+        row-major (C-style) or column-major (Fortran-style) order.
+    ndim : int, optional
+        Specifies the number of dimensions, i.e., the length of
+        ``dims``. This is required if the dimensions cannot be determined
+        from ``dims`` itself, for example, if ``dims`` is a Theano vector.
+
+    Returns
+    -------
+    unraveled_coords : tuple of ndarray
+        Each array in the tuple has the same shape as the ``indices``
+        array.
+
+    See Also
+    --------
+    ravel_multi_index
+
+    """
+    if ndim is None:
+        if isinstance(dims, (tuple, list)):
+            ndim = len(dims)
+        elif isinstance(dims, np.ndarray) and dims.ndim == 1:
+            ndim = dims.shape[0]
+        else:
+            raise TypeError('unravel_index was called with a dimension '
+                            'list with an unspecified length (dim = %s). '
+                            'Use the ndim parameter of unravel_index to '
+                            'set the number of dimensions. ' % str(dims))
+
+    res = UnravelIndex(order=order)(indices, dims, ndim)
+    if ndim == 1:
+        return (res,)
+    else:
+        return tuple(res)
+
+
+class RavelMultiIndex(gof.Op):
+    __props__ = ('mode', 'order')
+
+    def __init__(self, mode='raise', order='C'):
+        self.mode = mode
+        self.order = order
+
+    def make_node(self, *inp):
+        multi_index = [as_tensor_variable(i) for i in inp[:-1]]
+        dims = as_tensor_variable(inp[-1])
+
+        for i in multi_index:
+            if i.dtype not in int_dtypes:
+                raise TypeError("'%s' object cannot be interpreted as an index" % str(i.dtype))
+        if dims.dtype not in int_dtypes:
+            raise TypeError("'%s' object cannot be interpreted as an index" % str(dims.dtype))
+        if dims.ndim != 1:
+            raise TypeError("dims must be a 1D array")
+
+        return gof.Apply(
+            self, multi_index + [dims],
+            [TensorType(dtype=multi_index[0].dtype,
```

Comment:
from a quick test of numpy, the output dtype seem to be always int64 on my computer. It don't seem based on the input dtype.